synapse/types.py

```diff
@@ -138,24 +139,43 @@ def __deepcopy__(self, memo):
     @classmethod
     def from_string(cls, s):
         """Parse the string given by 's' into a structure object."""
-        if len(s) < 1 or s[0:1] != cls.SIGIL:
-            raise SynapseError(400, "Expected %s string to start with '%s'" % (
-                cls.__name__, cls.SIGIL,
-            ))
 
-        parts = s[1:].split(':', 1)
-        if len(parts) != 2:
+        if not s.startswith(cls.SIGIL):
+            raise SynapseError(
+                400, "Expected %s string to start with %r" % (
+                    cls.__name__, cls.SIGIL,
+                )
+            )
+
+        port = None
+        try:
+            localpart, domain = s[1:].split(':', 1)
+            if ':' in domain:
+                domain, port = domain.split(':')
```

Comment:
this looks like it will fail on IPv6 addresses.

Reply:
Good point! Completely forgot IPv6.